### PR TITLE
[182] handle project status cul-de-sacs

### PR DESCRIPTION
**Pull request by caravinci**
As per the ticket, this commit enables admins to revert project states that were basically "dead ends".
i.e. they can now "re-open" and "un-deny" projects, etc.

### CODE REVIEW

File: enrolment-service/enrol/core/views.py [hunk 1]
```diff
@@ -1,11 +1,11 @@
+# For email dev testing, use 'python -m smtpd -n -c DebuggingServer localhost:1025'
```
Comment (caravinci) [inline enrolment-service/enrol/core/views.py:1]: Little goodie to help dev time

File: enrolment-service/enrol/core/views.py [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 import os
-import json
 import logging
-from core.models import Applications, BillingContact, Projects, ProjectUsers
-from django.contrib.auth.models import User
+from core.models import Applications, Projects, ProjectUsers
 from django.conf import settings
 from django.db.models import Q
-from django.shortcuts import render, get_object_or_404
+from django.shortcuts import get_object_or_404
```
Comment (caravinci) [inline enrolment-service/enrol/core/views.py:8]: Pruning unused imports

File: enrolment-service/enrol/core/views.py [hunk 3]
```diff
@@ -191,39 +187,62 @@ def retrieve(self, request, pk=None):
...
+        project = self.get_object()
+        previous_status = project.status
+        new_status = serializer.validated_data['status']
+        # project_before = serializer.instance
+
+        logger.debug('updating id %s', project.id)
+        logger.debug('updating previous status %s', project.status)
+        logger.debug('updating incoming %s', serializer.validated_data)
+        logger.debug('updating by user %s', user)
+
```
Comment (caravinci) [inline enrolment-service/enrol/core/views.py:200]:
```suggestion
```

Comment (caravinci) [reply enrolment-service/enrol/core/views.py:200]: Forgot to remove these logs...

File: enrolment-service/enrol/core/views.py [hunk 4]
```diff
@@ -191,39 +187,62 @@ def retrieve(self, request, pk=None):
         return Response(status=status.HTTP_404_NOT_FOUND)
 
     def perform_update(self, serializer):
+        user = self.request.user
+        project = self.get_object()
+        previous_status = project.status
```
Comment (caravinci) [inline enrolment-service/enrol/core/views.py:192]: This helps checking what the current state flow intends, so we can send the right email notifications (e.g. newly accepted project, vs reopened... to be implemented in a different ticket)

File: enrolment-service/enrol/core/views.py [hunk 5]
```diff
@@ -191,39 +187,62 @@ def retrieve(self, request, pk=None):
...
+        logger.debug('updating id %s', project.id)
+        logger.debug('updating previous status %s', project.status)
+        logger.debug('updating incoming %s', serializer.validated_data)
+        logger.debug('updating by user %s', user)
+
         # Save the data
-        project = serializer.save()
+        updated_project = serializer.save()
+        project_users = ProjectUsers.objects.filter(project=project.id)
+        project_users.update(status=new_status)
```
Comment (caravinci) [inline enrolment-service/enrol/core/views.py:204]: D.R.Y

File: enrolment-service/enrol/core/views.py [hunk 6]
```diff
@@ -191,39 +187,62 @@ def retrieve(self, request, pk=None):
...
-                'message': 'The request from email <strong>{}</strong> to terminate project <strong>"{}"</strong> is now complete.'.format(
-                    project.user.email, project.project_name)
-            }
-            send_update_notification(email)
-            # print(email) # for testing
+        # Termination Confirmation
+        elif updated_project.status == 4:
+
+            # confirming a request.
+            if previous_status == 3:
```
Comment (caravinci) [inline enrolment-service/enrol/core/views.py:229]: Only use the existing "send a confirmation email" flow if the change is accepting a PI request.

File: enrolment-service/enrol/enrol/urls.py [hunk 7]
```diff
@@ -26,7 +26,7 @@
 projects_router = routers.NestedSimpleRouter(
     router, r'projects', lookup='project')
 projects_router.register(
-    r'users', views.ProjectUsersViewSet, base_name='project-users')
+    r'users', views.ProjectUsersViewSet, basename='project-users')
```
Comment (caravinci) [inline enrolment-service/enrol/enrol/urls.py:29]: Addresses a package API change

File: enrolment-service/enrol/yarn.lock [hunk 8]
```diff
@@ -1348,7 +1348,7 @@ object-visit@^1.0.0:
   dependencies:
```
Comment (caravinci) [inline enrolment-service/enrol/yarn.lock:1348]: lock stuff. nothing to see here.

File: enrolment-service/requirements.txt [hunk 9]
```diff
@@ -9,7 +9,7 @@ django-cors-headers==2.1.0
 django-nose==1.4.5
 django-rest-auth==0.9.1
 django-rest-swagger==2.1.0
-djangorestframework==3.9.1
+djangorestframework==3.11.2
```
Comment (caravinci) [inline enrolment-service/requirements.txt:12]: Updated to match the available package documentation

File: enrolment-ui/public/locales/en/translation.json [hunk 10]
```diff
@@ -44,13 +44,16 @@
...
             "deny": "Deny",
             "none": "No Action Needed",
             "pending": "Under Review",
             "purge": "Purge",
+            "rejectTermination": "Reject",
+            "reopen": "Reopen",
             "resendInvite": "Reinvite",
-            "terminate": "Terminate Project"
+            "reassess": "Reconsider",
+            "terminate": "Terminate"
```
Comment (caravinci) [inline enrolment-ui/public/locales/en/translation.json:56]: Adds translations, in English

File: enrolment-ui/public/locales/fr/translation.json [hunk 11]
```diff
@@ -44,13 +44,16 @@
...
             "none": "Aucune action nécessaire",
             "pending": "En cours de révision",
             "purge": "Effacer",
-            "resendInvite": "réinviter",
-            "terminate": "Résilier le projet"
+            "rejectTermination": "Refuser",
+            "reopen": "Reprendre",
+            "resendInvite": "Réinviter",
+            "reassess": "Réexaminer",
+            "terminate": "Résilier"
```
Comment (caravinci) [inline enrolment-ui/public/locales/fr/translation.json:56]: Adds translations, in French

File: enrolment-ui/src/index.js [hunk 12]
```diff
@@ -63,3 +64,8 @@ render(
   </Provider>,
   document.getElementById('root'),
 );
+
+// If you want to start measuring performance in your app, pass a function
+// to log results (for example: reportWebVitals(console.log))
+// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
+reportWebVitals(localStorage.DEBUG && console.log);
```
Comment (caravinci) [inline enrolment-ui/src/index.js:71]: Additional debugging tools

File: enrolment-ui/src/modules/Applications/ApplicationRequests.js [hunk 13]
```diff
@@ -91,56 +93,89 @@ const ApplicationRequests = ({
                                             <div className="requests-actionsColumn">
                                                 {application
                                                     ? (
```
Comment (caravinci) [inline enrolment-ui/src/modules/Applications/ApplicationRequests.js:95]: This adds the UI controls for the new functionality

File: enrolment-ui/src/modules/Projects/Projects.js [hunk 14]
```diff
@@ -14,7 +14,7 @@ import {
   uiSelectProject,
   uiSelectTab,
   toggleProjectTerminationModal,
-} from '../Projects/redux';
+} from './redux';
```
Comment (caravinci) [inline enrolment-ui/src/modules/Projects/Projects.js:17]: the target folder is in the same parent. not sure why they wrote it that way. ¯\_(ツ)_/¯

File: enrolment-ui/yarn.lock [hunk 15]
```diff
@@ -12105,6 +12105,11 @@ wbuf@^1.7.3:
   dependencies:
```
Comment (caravinci) [inline enrolment-ui/yarn.lock:12105]: more lock stuff, move along...

**Review by caravinci — COMMENTED** [review]
A few comments to assist reviewing

File: enrolment-service/enrol/core/views.py (hunk 6 above)
Comment (joneubank) [reply enrolment-service/enrol/core/views.py:229]: Useful code comments on these .status checks would be to declare what state each number is.

File: enrolment-service/enrol/enrol/urls.py (hunk 7 above)
Comment (joneubank) [reply enrolment-service/enrol/enrol/urls.py:29]: Nice catch. This kind of change is infuriating lol

File: enrolment-service/enrol/core/views.py (hunk 1 above)
Comment (joneubank) [reply enrolment-service/enrol/core/views.py:1]: Maybe add a note about this in the README for developers

Comment (caravinci) [reply enrolment-service/enrol/core/views.py:1]: good idea! will change that right now

File: enrolment-service/enrol/core/views.py (hunk 6 above)
Comment (caravinci) [reply enrolment-service/enrol/core/views.py:224]: Agreed! figured adding those throughout the file would be beyond scope, though.
These were originally defined in the model, and I'm just reusing the existing code.
I thought adding the intention declaratively in comments would suffice without scope creep:
e.g. "termination confirmation", "confirming a request"... but I guess I should make those a bit more verbose.

Let's see what I can do, without needing to go beyond the method definition.

File: enrolment-service/enrol/core/views.py (hunk 1 above)
Comment (caravinci) [reply enrolment-service/enrol/core/views.py]: moved...

File: README.md [hunk 16]
```diff
@@ -170,4 +170,7 @@ As mentioned before, Google API credentials must be created in order for the log
 
 ### Step 4 - Run
 
-At this point if you visit the front-end `http://localhost` you will see the login page, now ready to login and authorize DACO approved Gmail addresses as well as accepting internal logins using the Django admin login setup in the config
+At this point if you visit the front-end `http://localhost` you will see the login page, now ready to login and authorize DACO approved Gmail addresses as well as accepting internal logins using the Django admin login setup in the config.
+
+While developing, you are able to use 'python -m smtpd -n -c DebuggingServer localhost:1025' to mock an email server.
+This tool logs the email to terminal; useful to test and debug notifications, etc.
```
Comment (caravinci) [inline README.md:176]: moves the mock mail server to the readme... 
I've got a development-specific document in the making, but is not quite ready yet.

**Review by joneubank — APPROVED** [review]
This all looks fine. Thanks for the cleanup efforts.

**Pull request merged**